### ISSUE TITLE
Authoring guidelines: sample tests

### ISSUE THREAD

**Issue by hobovsky** [codewars/docs#146]
Breakdown of #139

Set of guidelines related to writing sample tests.

Proposed guidelines:
- stick to coding guidelines
- structured into properly described blocks, groups, categories, fixtures 
- not a replacement for a good description. Description should be enough to drive the user through the kata to its completion (maybe this point is redundant with "description guidelines" part? should it be repeated here?)
- there's no requirement on how many sample tests should be there. there can be many, but there can be just a few, or even none. Note that the more sample tests, the better the user experience will probably be, but it's up to author to decide.
- adding test cases should not be more difficult than writing a solution. Don't expect newbies to add test cases to an 8 kyu kata.
- they can be used as:
  - example of how user solution is called,
  - example of how to use testing framework
  - example of some requirements, edge cases
  - example of performance requirements, for example a single test case with large input
- Initial tests should compile correctly and it should be possible to run them.
- if description has any examples, they should be provided also as sample tests
- should be fixed, predictable and repeatable. No random.
- cases from sample tests should appear also in submission tests, possibly as a part of fixed tests.
- If possible, should not use parts from preloaded. Preloaded snippet is not visible to users and it's difficult to reproduce sample tests locally.

**Comment by Blind4Basics** [comment]
some random thougths. And some are rather ideas and might not really be appropriate...

---
notes on the points above:

> ~~not necessary, but encouraged~~

As a matter of fact, an issue is systematically opened when there are no sample tests, nowadays. ;)
So I'd just remove everything in the "if absent" part x)

> show example of a call to solution

This will encourage to put only 1 single tests => I'd remove/reword it.

> use them..., performance requirements

They are generally not in the sample tests (because most of the time they need the random tests), and if they _are_, I'd strongly recommand to actually comment out the related code, so that the user can run it if he wants, but he doesn't get straight the time out wall when using the sample tests (going that way, the sample tests are actually close to useless!)

---
additions/ideas (?):

* sample tests aren't there to replace a correct/good description (?)
* except in some very special cases, having just 1 or 3 sample tests is bad. Aim for a reasonable coverage of the inputs/outputs.
* all the tests present in the sample tests should also be present in the fixed tests of the full test suite (the reversed is not true)
* except in some rare situations: ___don't ever put randomness in the sample tests___ x)
*if in doubt about what to put or not in the sample tests, just put all fixed tests of the full test suite...

--- 

Introduction:

something about their goal being to allow the user to have a small but rather representative set of inputs/outputs so that he can quickly test is code, get an idea about how the tests are actually done, ...

**Comment by Blind4Basics** [comment]
note: I guess that the "interactions problems" between data structures that are in preloaded, sample tests and full tests suites (I mean, the risks of cheating about that) will be describe in some other place? ("kata authoring" section, I guess?)

**Comment by hobovsky** [comment]
> As a matter of fact, an issue is systematically opened when there are no sample tests, nowadays. ;)
> So I'd just remove everything in the "if absent" part x)

That's because many users, authors, and problems are inspired by/familiar with regular OJ/competitive programming sites, where you always get some fixed examples and answers. The meaning of TDD approach for solving kata was totally lost in history of CW, but I like it and I'd like ot restore it. It's also very similar to the way how things are done in the "industry". I basically like the idea when user is provided with some (almost) empty stub of tests, reference to testing framework, and from there they have to work their way through all (properly described!) requirements with red/green/refactor TDD cycles.

**Comment by Blind4Basics** [comment]
brrrruh.... xp

;)

**Comment by hobovsky** [comment]
You are right. But that's mostly because 99% of kata are actually not coding kata, but OJ challenges. You can see some leftovers of the idea in old kata created by Jake, where there's no random tests, and he even replied somewhere (cannot find now, but will search) that sample tests can be added by users driven by described requirements.

But I am not even sure if the idea still holds, but you actually gave me an idea: bring it back, and create a couple of real coding kata! People could learn TDD and how coding is done in RL, a very nice lesson IMO :)

I will try to put it somehow into guidelines. Maybe remove it as you suggested, or maybe divide it stronger, and explicitly mention "OJ style" or "Kata style"? I will see what I can came up with.

@kazk what is your opinion on TDD-driven coding kata? What do site owners think about the initial idea?

**Comment by Blind4Basics** [comment]
(I currently feel like I actually don't know what you're talking about... :o )

**Comment by error256** [comment]
I'm not sure yet if completely empty sample tests (that don't even compile) should be considered bad, but those that do pass, even if it's `assert(true)`, should be acceptable, even if possibly discouraged. There may be some common practices, but in the end it's up to kata authors to choose.

**Comment by kazk** [comment]
> @kazk what is your opinion on TDD-driven coding kata?

Not sure if I'm understanding the question. Do you mean to force "real" TDD experience by not providing tests? I don't think sample tests ruins TDD. It's still test first approach. You just didn't write the initial set of tests.

In my opinion, the sample tests are samples. Just enough for you to get started. No need for a good coverage unlike submission tests. But the author can choose to provide more tests if they want to, as long as they have written the description well and not using it as an excuse for underspecified task. OJs often just provides sample inputs and outputs as well, so I don't think OJ style (not sure what that means to be honest) or not matters. I don't really have a strong opinion about the amount of tests in the sample tests.

I'm strongly against sample tests that don't even compile/run though. This means the initial solution should be syntactically valid as well. For example:
```python
# BAD
def example(a, b):

# GOOD
def example(a, b):
    pass

def examle(a, b):
    return a
```

Users should be able to run the provided code and sample tests to see test failures. It's difficult for the runner to provide meaningful error when the tests are just broken.

I think it's difficult to define a strict rule about the amount of sample tests. I think most users would prefer more tests in sample tests and a good amount depends on the task as well. Also, many of the challenges on Codewars is underspecified and relies on the tests to reveal the requirements so that needs to be improved or users will be super frustrated.

We should definitely explain how it can be used while working on a solution. The flexibility of the sample tests (or being unit test based in general) is one of the major advantages of Codewars over other platforms. It's so much easier to work with than having to enter line separated values in specific format.

> What do site owners think about the initial idea?

I'm assuming "the initial idea" refers to not providing sample tests at all. We never do that at Qualified. Even when we specifically want to see the candidate's ability to write good tests, we provide few valid tests, and clearly mention in the instruction that they're expected to write their own tests while solving. We generally try to optimize the candidate's experience as much as we can and have specific goals for each challenge.

Requiring users to write tests makes sense if you're testing/teaching that, but it can become a noise if that's not the main focus. On Codewars, how you solve a kata doesn't really matter. You can do TDD, but no one knows and it's not part of the challenge.

As I mentioned above, I don't think the idea of TDD is lost with sample tests. Having no sample tests at all or broken sample tests just increases unnecessary friction.

@jhoffner anything to add?

I'm also inviting @ggorlen. Greg makes challenges at Qualified. Challenges on Codewars and Qualified can have different focus, but many of the best practices at Qualified should be relevant to Codewars.

**Comment by hobovsky** [comment]
Okay so it seems I got things a bit backwards then. Or maybe did not understand them well and built some different image in my mind.

What I **thought**: 

Kata is an excercise where you start with little to none code (both tests and solution are almost empty) and using reed/green/refractor cycles you go through requirements, add failing cases, add implementation, rinse, repeat. When cannot think of more test cases, or when you consider solution ready, you submit and see what happens. If you fail, you add another test case, implement it, and so on. That's what I meant by "tdd style" and that's what I thought the original idea could be, because that's how I thought coding kata work. Apparently, I was wrong. 

Later when authors with OJ experience appeared and started creating tasks, i thought that, since they might be not familiar with kata and tdd, but familiar with algorithmic challenges and input and output examples, they created sample tests to mimic example input and output. And if they passed, they were either representative enough to make full test suite pass, or caused a lot of head scratching, guessing and frustration about how to continue if full tests failed. And this part I also probably got wrong :)

It seems that I simply misunderstood the topic. I will rethink it and try to come up with something clearer. 

Did I make other parts also so terribly twisted? :)

**Comment by kazk** [comment]
I don't know what the original idea was exactly. Jake might comment on that.

We're probably just misunderstanding something. I don't see much difference between your TDD style and OJ style. The only difference is the mindset of the solver. More specifically, what the solver is expecting from the sample tests.

In TDD style, the solver knows they're not shown all the possible cases. They have and are prepared to write their own tests.
In OJ style, the solver is not sure if passing sample tests is enough. But they _might_ have written own tests while solving.

Even in TDD style, the description will provide some examples, right? The sample tests should at least include them. I wouldn't say this is influenced by authors with OJ experience. It's just better user experience because almost everyone will add those tests when starting.

OJs don't give you all the possible inputs and outputs including edge cases. I don't think our sample tests should either. The sample tests shouldn't be equivalent to the submission tests. It's there for the solvers to use when needed.

This is why I've been saying I don't know what you mean by OJ style.

**Comment by hobovsky** [comment]
> We're probably just misunderstanding something. I don't see much difference between your TDD style and OJ style. The only difference is the mindset of the solver. More specifically, what the solver is expecting from the sample tests.
>
> In TDD style, the solver knows they're not shown all the possible cases. They have written and are prepared to write their own tests.
> In OJ style, the solver is not sure if passing sample tests is enough. But they might have written own tests while solving.

Yes, this part is almost exactly what I meant. And it comes from following reasoning:
- I think that many users are not aware they can, need, or should add sample tests. But now I see it should probably be explained with documentation.
- Sample tests often have to make up for underspecified descriptions. But now I see it should be handled by fixing descriptions and proper review.
- Feedback from submission tests is usually too poor to provide hints on what sample test to add. But it should be fixed by giving better feedback from assertions and full tests.
- I believe that any users expect the sample tests to be representative enough to let you pass the kata. That's why you get so many complaints that "all tests pass but I cannot submit, whyy??!?". I do not know how to fix this.
- I believe that current, unwritten rules require sample tests to be representative enough because otherwise the above issues start to pile up. It probably would not matter that much if descriptions were good, and tests suites provided better feedback.

**Comment by ggorlen** [comment]
I think the latest adjustment is a good improvement. Separate "TDD" and "OJ" (online judge, I assume?) styles seem to be a false dichotomy and/or oversimplification. Many challenge styles are possible and I assume the guidelines are looking to focus on what the community sees as universal best practice (or as close to universal as possible) agnostic of style.

For example, the community standard is that if the sample tests crash out of the box and the description doesn't explicitly state this to be part of the challenge, it's a bug that harms user experience and will draw complaints/low ratings. Basically, if any of these guidelines are broken, the description should justify it with a plausible explanation. This is all part of ensuring the challenger knows exactly what/where the goalpost is so they can kick the ball through it, so to speak.

One example of breaking the rules correctly is the [Thinking & Testing](https://www.codewars.com/collections/thinking-and-testing) series. The description justifies why it breaks a fundamental rule and the value proposition/intent is obvious.

---

Other suggestions:

> should be fixed, predictable and repeatable. No random.

I'd expand "fixed/predictable/repeatable" to recommend that each test case be pure and not rely on state from other tests. This is implied by the existing wording, but it's subtle and it might not be obvious that creating one instance of the candidate's solution class and reusing it across multiple test cases could be a problem. I've fallen into this in the past; on some challenge types, it takes extra thinking and work to destroy and rebuild state.

> Preloaded snippet is not visible to users

Preloaded is not directly visible, but this file is readable programmatically from the solution code, so only non-sensitive things should be put into it like boilerplate that downloads a CSV or opens a DB connection. Some kata are solvable by reading the contents of preloaded when the author didn't realize it wasn't invisible.

> structured into properly described blocks, groups, categories, fixtures

This could be expanded to state that there should be only a couple of related assertions per test case block.

Bad:
```js
it("should work on basic functionality", () => {
  // 10 different assertions testing basic functionality
});
```
Good:
```js
it("should add two numbers", () => {
  expect(solution(...)).toEqual(...);
});
it("should frob the foobar when quux is garply", () => {
  expect(solution(...)).toFrobTheFoobar.when.quux.isGarply();
});
// .... 8 more test cases that test something unique and specific
```

Piggybacking on that, I'd add something about test names and assertion error messages here. Descriptive test names go a long way to providing useful feedback to the candidate in terms of what went wrong. `it("should work"...` vs `it("should make a GET request to foo/bar and return 200"...`. For the assertion failure, default messages are [sometimes unhelpful](https://www.qualified.io/kb/languages/c/criterion#verbose-output-for-complex-data) like `"assertion failed"` which should be overridden with, say, `"expected response status 200 but got 404"`. Catch promise errors that might time out or cause confusing messages in the test suite, close DB connections on sad paths, etc.

Although this applies more to hidden tests, I believe it's good practice to avoid redundant tests. Include the right amount needed to validate the specification. Some authors tend to use redundant tests to create the illusion of good coverage or use loops to generate hundreds or thousands of hidden random tests that are used to enforce the specification and test efficiency. Introducing nondeterminism can result in solutions that occasionally pass and occasionally fail which is bad news. Coverage aside, _why_ a random test failed among thousands can be obscure for the user because it has no name. There only needs to be just enough random tests to disqualify hardcoded solutions. Performance, core specification and edge cases should be deterministic named tests covering disjoint and localizable/specific requirements.

Following these guidelines should help reduce the "I think my code works but the submission suite gives me an error and I don't know why" problem, among others expressed in the above post. Writing katas is difficult and there are so many holes to plug so it's not surprising to see these remarks pop up.

Concrete examples of violations of these guidelines might help illustrate and I'd be happy to provide some if this would be useful.

We have a list of dozens of best practices at Qualified and maybe we can open those up to CW docs--I'm sure they'd benefit from the community feedback and most of the fundamental guidelines should be the same agnostic of platform so we can save effort on both sides.

I also see there are many different issues/topics open and I apologize if some of this was covered elsewhere or might overlap with a different topic.

**Comment by hobovsky** [comment]
Most of your remark would fit better for #19 and #151, i will try to distribute them somehow.

**Comment by ggorlen** [comment]
Ah, I see that thread. Most of my suggestions were already mentioned there. Thanks.

**Comment by hobovsky** [comment]
@ggorlen now you've seen that part, would you please be so kind and remove from above post parts which you think are addressed there sufficiently, and copy /paste parts you'd like to have improved? It would help me a lot to bring it closer to your expectations. I think most of your remarks belong to Fixed tests part of the full test suite description rather than here.

I would really appreciate. 

Thank you!

**Comment by hobovsky** [comment]
Closed with #175
